business/proxy_status.go

```diff
@@ -0,0 +1,129 @@
+package business
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/kiali/kiali/kubernetes"
+	"github.com/kiali/kiali/models"
+)
+
+type ProxyStatus struct {
+	k8s kubernetes.ClientInterface
+}
+
+func (in *ProxyStatus) GetWorkloadProxyStatus(workloadName, namespace string, workloadStatus models.WorkloadStatus) models.WorkloadStatus {
+	proxyStatuses, err := in.k8s.GetProxyStatus()
+	if err != nil {
+		return workloadStatus
+	}
+
+	ps := getProxyStatusByName(workloadName, namespace, proxyStatuses)
+	if ps != nil {
+		workloadStatus.ProxyStatus = castProxyStatus(*ps)
+	}
+	return workloadStatus
+}
+
+func (in *ProxyStatus) GetWorkloadsProxyStatuses(namespace string, workloadStatuses []models.WorkloadStatus) []models.WorkloadStatus {
+	res := make([]models.WorkloadStatus, 0, len(workloadStatuses))
+
+	proxyStatuses, err := in.k8s.GetProxyStatus()
+	if err != nil {
+		return workloadStatuses
+	}
+
+	for _, ws := range workloadStatuses {
+		ps := getProxyStatusByName(ws.Name, namespace, proxyStatuses)
+		if ps != nil {
+			ws.ProxyStatus = castProxyStatus(*ps)
+		}
+		res = append(res, ws)
+	}
+
+	return res
+}
+
+func (in *ProxyStatus) GetNamespaceAppProxyStatus(namespace string, appsWorkloads models.NamespaceAppHealth) (models.NamespaceAppHealth, error) {
+	proxyStatuses, err := in.k8s.GetProxyStatus()
+	if err != nil {
+		return appsWorkloads, err
+	}
+
+	for app, workloadStatus := range appsWorkloads {
+		workloadStatuses := make([]models.WorkloadStatus, 0, len(workloadStatus.WorkloadStatuses))
+
+		for _, ws := range workloadStatus.WorkloadStatuses {
+			ps := getProxyStatusByName(ws.Name, namespace, proxyStatuses)
+			if ps != nil {
+				ws.ProxyStatus = castProxyStatus(*ps)
+			}
+			workloadStatuses = append(workloadStatuses, ws)
+		}
+		appsWorkloads[app].WorkloadStatuses = workloadStatuses
+	}
+
+	return appsWorkloads, nil
+}
+
+func (in *ProxyStatus) GetNamespaceWorkloadProxyStatus(namespace string, workloads models.NamespaceWorkloadHealth) (models.NamespaceWorkloadHealth, error) {
+	res := models.NamespaceWorkloadHealth{}
+
+	proxyStatuses, err := in.k8s.GetProxyStatus()
+	if err != nil {
+		return workloads, err
+	}
+
+	for wl, ws := range workloads {
+		ps := getProxyStatusByName(wl, namespace, proxyStatuses)
+		if ps != nil {
+			ws.WorkloadStatus.ProxyStatus = castProxyStatus(*ps)
+		}
+		res[wl] = ws
+	}
+
+	return res, nil
+}
+
+func getProxyStatusByName(name, namespace string, proxyStatus []*kubernetes.ProxyStatus) *kubernetes.ProxyStatus {
+	for _, ps := range proxyStatus {
+		if strings.HasPrefix(ps.ProxyID, name) && strings.HasSuffix(ps.ProxyID, namespace) {
+			return ps
+		}
+	}
+	return nil
+}
+
+func castProxyStatus(ps kubernetes.ProxyStatus) []models.ProxyStatus {
+	statuses := make([]models.ProxyStatus, 0, 4)
+
+	r := reflect.ValueOf(ps)
```

Comment:
I'm not sure why reflection is needed here; aren't those fields explicitly defined in `istio.go` `SyncStatus` ? Why can't we have direct/explicit access on those?

(btw if we have to go through reflection I think some defensive checks are necessary, in the case fields don't exist)

Reply:
Yes, the fields are specified in the SyncStatus. However, I wanted to avoid the repetition of performing the same action with every pair of fields.

```
if xdsStatus(ps.ClusterSent, ps.ClusterAcked) != models.Synced {
  statuses = append(statuses, models.ProxyStatus{})
}

if xdsStatus(ps.EndpointSent, ps.EndpointAcked) != models.Synced {
  statuses = append(statuses, models.ProxyStatus{})
}

if xdsStatus(ps.ListenerSent, ps.ListenerAcked) != models.Synced {
  statuses = append(statuses, models.ProxyStatus{})
}


if xdsStatus(ps.RouteSent, ps.RouteAcked) != models.Synced {
  statuses = append(statuses, models.ProxyStatus{})
}
```

Is there any other way to perform this differently?
How would you have approached that?

Reply:
I see, perhaps it wouldn't be as much simplified as you'd like, but maybe with some temporary struct?

Something like this:

```go

	type cpntStatus struct {
		name string
		sent string
		ack  string
	}
	cluster := cpntStatus{name: "CDS", sent: ps.ClusterSent, ack: ps.ClusterAcked}
	endpoint := cpntStatus{name: "EDS", sent: ps.EndpointSent, ack: ps.EndpointAcked}
	listener := cpntStatus{name: "LDS", sent: ps.ListenerSent, ack: ps.ListenerAcked}
	route := cpntStatus{name: "RDS", sent: ps.RouteSent, ack: ps.RouteAcked}
	for _, cpnt := range []cpntStatus{cluster, endpoint, listener, route} {
// etc.
```

But to be honest, personally I probably wouldn't have cared much about this repetition (at least as long as it's not very big). Reflection has a cost not only in performances but also in maintenance, being more error-prone (e.g. if in the future we change the name of some fields in the underlying struct: reflection we miss the help of the compiler to detect where they're used,  making refactorings harder and more risky). Surely if we had many more repetitions, the pros of reflection would surpass the cons.

Reply:
Could be condensed:

```go
	type cpntStatus struct {
		name, sent, ack string
	}
	for _, cpnt := range []cpntStatus{{"CDS", ps.ClusterSent, ps.ClusterAcked}, {"EDS", ps.EndpointSent, ps.EndpointAcked}, {"LDS", ps.ListenerSent, ps.ListenerAcked}, {"RDS", ps.RouteSent, ps.RouteAcked}} {
```